icu4c/source/common/unicode/localpointer.h

```diff
@@ -42,7 +42,11 @@
 
 #if U_SHOW_CPLUSPLUS_API
 
+#ifndef U_IN_DOXYGEN
+extern "C++" {
```

Comment:
That violates the C++ standard, I think, as C++17 [using.headers]/3 states: "A translation unit shall include a header only outside of any declaration or definition..." (and an extern "C++" {...} linkage-specification is a declaration).

Reply:
Hi,
Notice that you have this also here[1].
This what allowed libraries to include the headers files also in C context so far as you explicitly entered C++ context.
The recent addition of the `#include <memory>` without explicit enter C++ context broke some consumers of ICU.
This is actually a regression in the last release.
I do not think you should have mixed C and C++ headers, a C application should include C code, regardless if the compiler is C++... this is also how C include files are managed. For some reason in ICU there is a mix of the two, and I guess this why the solution of entering explicit C++ context was introduced.
However, as long as this is the case, I suggest to add this explicit C++ scope as well and later on think how to split the C++ headers so that they will not implicitly included in C application.
What do you think?
Thanks!

[1] https://github.com/unicode-org/icu/blob/master/icu4c/source/common/unicode/uversion.h#L119

Reply:
> Notice that you have this also here[1].

That's
```
#   define U_NAMESPACE_BEGIN extern "C++" { namespace U_ICU_NAMESPACE {
# define U_NAMESPACE_END } }
```
which in itself doesn't cause that problem.  Only if you had an #include for a system header between U_NAMESPACE_BEGIN and U_NAMESPACE_END, then that would cause the problem I mentioned.

> This what allowed libraries to include the headers files also in C context so far as you explicitly entered C++ context.

With "C context" and "C++ context" you mean extern "C" resp. extern "C++" linkage-specifications?  I think it is a broken approach to allow client code to #include an ICU include file nested from within such a linkage-specification.  (But I know little about ICU.  I only got prompted to this issue through a report of a failed LibreOffice build; mail thread starting at <https://lists.freedesktop.org/archives/libreoffice/2019-March/082340.html> "Build failure with latest ICU 64.1".)

Reply:
Correct, it is *not* the problem, but in normal cases the global scope should be assumed as C++. The fact that ICU had this allowed maybe subjective and not best-practice invalid usage so far. In 64.1 there is a regression of introducing C++ code without entering the context, so these non-best practices that used ICU were broken.

I do not know why ICU had this `extern "C++"` it should probably be removed, then ICU has a consistent expectation of being in global C++ context.

There is a regression in ICU, even if we argue that the fault is at the "other side", the regression was introduced by mistake without understanding the implication while adding additional include in global context as part of performing other activity.

If the assumption of global context is now C++, a different patch should be added to clearly note that as poential breakage and remove the explicit `extern "C++"` as it is useless anyway.

Reply:
hmmm.... while trying to write the alternative of removing the `extern "C++"` I found the intention.

Look at ICU own documentation[1]... the policy was so far explicitly to enable include within extern "C" scopes:

```
/**
 * \def U_NAMESPACE_BEGIN
 * This is used to begin a declaration of a public ICU C++ API.
 * When not compiling for C++, it does nothing.
 * When compiling for C++, it begins an extern "C++" linkage block (to protect
 * against cases in which an external client includes ICU header files inside
 * an extern "C" linkage block).
 *
 * It also begins a versioned-ICU-namespace block.
 * @stable ICU 2.4
 */
```

[1] https://github.com/unicode-org/icu/blob/master/icu4c/source/common/unicode/uversion.h#L61

Reply:
However, I do see that the violation began to spread...
```
char16ptr.h:#include <cstddef>
localpointer.h:#include <memory>
std_string.h:#include <string>
unistr.h:#include <cstddef>
```

Reply:
> hmmm.... while trying to write the alternative of removing the `extern "C++"` I found the intention.
> 
> Look at ICU own documentation[1]... the policy was so far explicitly to enable include within extern "C" scopes:
> 
> ```
> /**
>  * \def U_NAMESPACE_BEGIN
>  * This is used to begin a declaration of a public ICU C++ API.
>  * When not compiling for C++, it does nothing.
>  * When compiling for C++, it begins an extern "C++" linkage block (to protect
>  * against cases in which an external client includes ICU header files inside
>  * an extern "C" linkage block).

The important thing which you missed in your quote is that `U_NAMESPACE_BEGIN` also contains `namespace U_ICU_NAMESPACE {` which means that it would never be allowed around `#include` lines. Its purpose was only to be placed around declarations of C++ classes.

Reply:
The mission statement in this comment is to allow including the icu headers from extern "C", nobody suggested that the same macro will be used, in time some C++ headers were added in global scope, breaking this assumption. Either upstream wishes to keep this requirement and apply this fix or I will submit a different patch of removing the requirement.

Please from now on, let upstream comment.